src/v/cluster/security_frontend.cc

```diff
@@ -323,23 +323,41 @@ security_frontend::get_bootstrap_user_creds_from_env() {
     }
 
     ss::sstring creds_str = creds_str_ptr;
-    auto colon = creds_str.find(":");
-    if (colon == ss::sstring::npos || colon == creds_str.size() - 1) {
+    vector<size_t> colon_positions;
+    size_t pos = creds_str.find(sub, 0);
+    while (pos != ss::sstring::npos) {
+        positions.push_back(pos);
+        pos = creds_str.find(sub, pos + 1);
+    }
+    if (
+      colon_positions.size() != 2
+      || colon_positions.back() == creds_str.size() - 1) {
         // Malformed value.  Do not log the value, it may be malformed
         // but it is still a secret.
         vlog(
           clusterlog.warn,
-          "Invalid value of {} (expected \"username:password\")",
+          "Invalid value of {} (expected \"username:password:mechanism\")",
           bootstrap_user_env_key);
         return {};
     }
-
-    auto username = security::credential_user{creds_str.substr(0, colon)};
-    auto password = creds_str.substr(colon + 1);
-    auto credentials = security::scram_sha256::make_credentials(
-      password, security::scram_sha256::min_iterations);
-    return std::optional<user_and_credential>(
-      std::in_place, std::move(username), std::move(credentials));
+    auto username = security::credential_user{
+      creds_str.substr(0, colon_positions.front())};
+    auto password = creds_str.substr(
+      colon_positions.front() + 1, colon_positions.back());
+    auto provided_mechanism = creds_str.substr(colon_positions.back() + 1);
+    if (provided_mechanism == "SCRAM-SHA-512") {
+        auto credentials = security::scram_sha512::make_credentials(
+          password, security::scram_sha512::min_iterations);
+
+        return std::optional<user_and_credential>(
+          std::in_place, std::move(username), std::move(credentials));
+    } else {
+        auto credentials = security::scram_sha256::make_credentials(
+          password, security::scram_sha256::min_iterations);
+
+        return std::optional<user_and_credential>(
+          std::in_place, std::move(username), std::move(credentials));
+    }
```

Comment:
This version:
* Cleans up the code a bit
* Rejects empty username or password
* Selects SCRAM-SHA-512 if and only if there is a 3rd part that matches the mechanism name
* Is backwards compatible

I still think it should reject an invalid mechanism.

```suggestion
    std::string_view creds = creds_str_ptr;
    std::vector<std::string_view> parts;
    parts.reserve(3);
    boost::algorithm::split(parts, creds, [](char c) { return c == ':'; });
    if (
      parts.size() != 2 || parts.size() != 3 || parts[0].empty()
      || parts[1].empty()) {
        // Malformed value.  Do not log the value, it may be malformed
        // but it is still a secret.
        vlog(
          clusterlog.warn,
          "Invalid value of {} (expected \"username:password:mechanism\")",
          bootstrap_user_env_key);
        return {};
    }
    using make_credentials_t = security::scram_credential (*)(
      const ss::sstring&, int);
    make_credentials_t make_credentials
      = &security::scram_sha256::make_credentials;
    if (
      parts.size() == 3
      && parts[2] == security::scram_sha512_authenticator::name) {
        make_credentials = &security::scram_sha512::make_credentials;
    }

    auto credentials = make_credentials(
      ss::sstring{parts[1]}, security::scram_sha512::min_iterations);

    return std::optional<user_and_credential>(
      std::in_place,
      security::credential_user{parts[0]},
      std::move(credentials));
```

Reply:
Thanks for the changes Ben, especially around ensuring backwards compatibility! I've pushed this to the branch.